custom/inddex/reports/r2b_gaps_detail.py

```diff
@@ -47,40 +47,34 @@ def data_providers(self):
         ]
 
 
-def _matches_filters(row, selected_gap_type, selected_gap_code):
-    # If a gap type is specified, show only rows with gaps of that type
-    gap_type = selected_gap_type
-    if gap_type == ConvFactorGaps.slug and row.conv_factor_gap_code == ConvFactorGaps.AVAILABLE:
-        return False
-    if gap_type == FctGaps.slug and row.fct_gap_code == FctGaps.AVAILABLE:
-        return False
-
-    # gap_code is from a drilldown filter, so gap_type must also be selected
-    if gap_type and selected_gap_code:
-        if gap_type == ConvFactorGaps.slug and str(row.conv_factor_gap_code) != selected_gap_code:
-            return False
-        if gap_type == FctGaps.slug and str(row.fct_gap_code) != selected_gap_code:
+def _matches_filters(row, gap_class, gap_code, selected_gap_type, selected_gap_code):
+    if selected_gap_type:
+        if gap_class.slug != selected_gap_type:
             return False
 
+        if selected_gap_code:
+            gap_code = str(row.conv_factor_gap_code if selected_gap_type == ConvFactorGaps.slug
+                           else row.fct_gap_code)
+            return gap_code == selected_gap_code
+
     return True
 
 
 def get_gaps_data(food_data, selected_gap_type=None, selected_gap_code=None):
     for row in food_data.rows:
-        if _matches_filters(row, selected_gap_type, selected_gap_code):
-            for gap_class, gap_code in [
-                    (ConvFactorGaps, row.conv_factor_gap_code),
-                    (FctGaps, row.fct_gap_code),
-            ]:
-                if not selected_gap_type or selected_gap_type == gap_class.slug:
-                    manually_set = ['gap_type', 'gap_code', 'gap_desc', 'number_occurrence']
-                    yield {
-                        'gap_type': gap_class.name,
-                        'gap_code': gap_code,
-                        'gap_desc': gap_class.DESCRIPTIONS[gap_code],
-                        **{col: getattr(row, col, None) for col in GapsDetailsData.headers
-                        if col not in manually_set},
-                    }
+        for gap_class, gap_code in [
+                (ConvFactorGaps, row.conv_factor_gap_code),
+                (FctGaps, row.fct_gap_code),
+        ]:
+            if _matches_filters(row, gap_class, gap_code, selected_gap_type, selected_gap_code):
+                manually_set = ['gap_type', 'gap_code', 'gap_desc', 'number_occurrence']
+                yield {
+                    'gap_type': gap_class.name,
+                    'gap_code': gap_code,
+                    'gap_desc': gap_class.DESCRIPTIONS[gap_code],
+                    **{col: getattr(row, col, None) for col in GapsDetailsData.headers
+                    if col not in manually_set},
```

Comment:
In this case, irrelevant, but [useful to know](https://wiki.python.org/moin/TimeComplexity):
* `thing in list_of_things`: average: O(n)
* `thing in set_of_things`: average: O(1), worst O(n)

Reply:
I was actually thinking about that a bit, but I know `set` has some tradeoffs of it's own, like a greater overhead.  Set definitely wins out with large collections and frequent references, but with small collections and frequent creations (I should move `manually_set` higher up) I think it's more of a toss-up.  Then I caught myself bikeshedding and moved on :stuck_out_tongue:

Reply:
Since this sort of question is [https://xkcd.com/356/](irresistible)...
```python
In [1]: from timeit import timeit

In [2]: def do_list():
   ...:     a = [1, 2, 3, 4, 5]
   ...:     return 6 in a
   ...: 

In [3]: def do_set():
   ...:     a = {1, 2, 3, 4, 5}
   ...:     return 6 in a
   ...: 

In [4]: timeit(do_list)
Out[4]: 0.19533048500306904

In [5]: timeit(do_set)
Out[5]: 0.20172523200744763
```
And if the member _is_ present, `do_list` goes down to about 0.15, whereas `do_set` stays the same.  With the collection defined outside the function, `do_set` comes out slightly ahead.

Reply:
That is really interesting! Thank you!

Reply:
> irresistible

:rofl: